python/tvm/relax/training/trainer.py

```diff
@@ -0,0 +1,252 @@
+        self._predict_fn = setup_trainer_pass.PREDICT_FUNC_NAME
+        self._adjoint_fn = setup_trainer_pass.ADJOINT_FUNC_NAME
+        self._update_params_fn = setup_trainer_pass.UPDATE_PARAMS_FUNC_NAME
+
+        # Compile & Build
+        backbone[self._predict_fn] = backbone[self._predict_fn].with_attr(
+            setup_trainer_pass.PARAMS_NUM_ATTR_KEY, params_num
+        )
+        self._mod = setup_trainer_pass(backbone)  # type: ignore
+        self._vm: Optional[VirtualMachine] = None
+        self._optim_state = self._mod[self._update_params_fn].attrs[
+            setup_trainer_pass.OPTIM_STATE_ATTR_KEY
+        ]
+
+        # Allocate Parameters
+        self._parameters_buffer: List[NDArray] = []
+        self._parameters_name_to_pos: Dict[str, int] = {}
+
+        def _convert_from_tvm_shape(tvm_shape):
+            return [int(dim) for dim in tvm_shape]
+
+        self._parameters_buffer = []
+        for i, param in enumerate(self._mod[setup_trainer_pass.ADJOINT_FUNC_NAME].params):
+            if i < params_num:
+                self._parameters_buffer.append(
+                    tvm.nd.array(
+                        np.zeros(
+                            shape=_convert_from_tvm_shape(param.struct_info.shape),
+                            dtype=np.dtype(param.struct_info.dtype),
+                        )
+                    )
+                )
+                self._parameters_name_to_pos[param.name_hint] = i
+
+    def build(
+        self,
+        target: Union[str, tvm.target.Target],
+        device: Union[tvm.runtime.Device, List[tvm.runtime.Device]] = tvm.cpu(0),
+        memory_cfg: Optional[str] = None,
+    ):
+        """Build and compile the module.
+
+        Parameters
+        ----------
+        target : Union[str, tvm.target.Target]
+            The target to run all modules on.
+
+        device : Union[Device, List[Device]]
+            The device, or devices on which to execute the VM code.
+
+        memory_cfg : Optional[str]
+            The allocator behavior to use for the VM.
+        """
+        ex = build(self._mod, target=target)
+        self._vm = VirtualMachine(ex, device=device, memory_cfg=memory_cfg)
+
+    def _check_build(self):
+        if self._vm is None:
+            raise TVMError("Please build the trainer first. Use `trainer.build(...)`.")
+
+    @property
+    def mod(self) -> IRModule:
+        """Return the IRModule transformed by the setup trainer pass.
+
+        Returns
+        -------
+        ret : IRModule
+            The IRModule transformed by the setup trainer pass.
+        """
+        return self._mod
+
+    @property
+    def vm(self) -> VirtualMachine:
+        """Return the relax virtual machine of the module. Should first build the trainer.
+
+        Returns
+        -------
+        ret : VirtualMachine
+            The relax virtual machine.
+        """
+        self._check_build()
+        return self._vm
+
+    def rand_init_params(self):
```

Comment:
Considering other initialiaztion methods, shall we use sth like `xavier_uniform_init_params` as name, just as torch does?